File: Programming/Practice/LeetCode/Arrays101/DuplicateZeros/solution.py
from typing import List


class Solution:
    def duplicateZeros(self, arr: List[int]) -> None:
        to_insert = [idx for idx, element in enumerate(arr) if element == 0]
        for offset, index in enumerate(to_insert):
            arr.insert(index + offset, 0)
        del arr[-len(to_insert):]

if __name__ == "__main__":
    array = [1,0,2,3,0,4,5,0]
    
    solution = Solution()
    solution.duplicateZeros(array)

    assert(
        array == [1,0,0,2,3,0,0,4]
    )

File: Programming/Practice/LeetCode/Arrays101/DuplicateZeros/solution.cpp
#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    void duplicateZeros(vector<int>& arr) {
        std::vector<int> to_insert;
        int size = arr.size();
        
        for(int i = 0; i < arr.size(); ++i) {
            if(arr[i] == 0) {
                to_insert.push_back(i);
            }
        }
        for(int i = 0; i < to_insert.size(); ++i) {
            auto it = arr.begin();
            std::advance(it, to_insert[i] + i);
            arr.insert(it, 0);
        }
        arr.resize(size);
    }
};

int main() {
    Solution solution = Solution();
    vector<int> sample_vector = {1,0,2,3,0,4,5,0};

    solution.duplicateZeros(sample_vector);

    for(auto it = sample_vector.begin(); it != sample_vector.end(); ++it) {
        cout << *it;
    }
}